sql/alps_submitter.go

```diff
@@ -0,0 +1,167 @@
+	Boundaries   []int
+}
+
+type CrossColumn struct {
+	Keys           []interface{}
+	HashBucketSize int
+}
+
+func resolveFeatureSpec(el *exprlist) (*FeatureSpec, error) {
+	headExpr := (*el)[0].val
+	if headExpr == "DENSE" {
+		shape, err := strconv.Atoi((*el)[2].val)
+		if err != nil {
+			return nil, err
+		}
+		return &FeatureSpec{
+			FeatureName: (*el)[1].val,
+			IsSparse:    false,
+			Shape:       []int{shape},
+			DType:       "double"}, nil
+	} else if headExpr == "SPARSE" {
+		shape, err := strconv.Atoi((*el)[2].val)
+		if err != nil {
+			return nil, err
+		}
+		return &FeatureSpec{
+			FeatureName: (*el)[1].val,
+			IsSparse:    false,
+			Shape:       []int{shape},
+			DType:       "double"}, nil
+	}
+	return nil, fmt.Errorf("not supported encoding type: %s", headExpr)
+}
+
+func resolveExpr(e *expr) (string, *exprlist) {
+	if e.val != "" {
+		return e.val, nil
+	} else {
+		return "", &e.sexp
+	}
+}
+
+func resolveExprList(el *exprlist) (interface{}, error) {
+	headName, headExprList := resolveExpr((*el)[0])
+	if headName == "" {
+		return resolveExprList(headExprList)
+	} else {
+		switch headName {
+		case "DENSE":
+			return resolveFeatureSpec(el)
+		case "SPARSE":
+			return resolveFeatureSpec(el)
+		case "NUMERIC":
+			shape, err := strconv.Atoi((*el)[2].val)
+			if err != nil {
+				return nil, err
+			}
+			key, subEl := resolveExpr((*el)[1])
+			if subEl != nil {
+				return nil, fmt.Errorf("key of NUMERIC must be string")
+			}
+			return &NumericColumn{
+				Key:   key,
+				Shape: shape}, nil
+		case "BUCKET":
+			sourceExprList := (*el)[1].sexp
+			boundariesExprList := (*el)[2].sexp
+			source, err := resolveExprList(&sourceExprList)
+			if err != nil {
+				return nil, err
+			}
+			boundaries, err := resolveExprList(&boundariesExprList)
+			if err != nil {
+				return nil, err
+			}
+			return &BucketColumn{
+				SourceColumn: *source.(*NumericColumn),
+				Boundaries:   transformToIntList(boundaries.([]interface{}))}, nil
+		case "CROSS":
+			keysExpr := (*el)[1].sexp
+			bucketSize, _ := strconv.Atoi((*el)[2].val)
+			keys, err := resolveExprList(&keysExpr)
+			if err != nil {
+				return nil, err
+			}
+			return &CrossColumn{
+				Keys:           keys.([]interface{}),
+				HashBucketSize: bucketSize}, nil
+		case "square":
+			var list []interface{}
+			for idx, expr := range *el {
+				if idx > 0 {
+					val, sexp := resolveExpr(expr)
+					if sexp == nil {
+						val, _ := strconv.Atoi(val)
+						list = append(list, val)
+					} else {
+						value, err := resolveExprList(sexp)
+						if err != nil {
+							return nil, err
+						} else {
+							list = append(list, value)
+						}
+					}
+				}
+			}
+			return list, nil
+		default:
+			return nil, fmt.Errorf("not supported expr in ALPS submitter: %s", headName)
+		}
+	}
+	return nil, nil
+}
+
+func transformToIntList(interfaceList []interface{}) []int {
+	var b []int
+	for i := range interfaceList {
+		b = append(b, interfaceList[i].(int))
+	}
+	return b
```

Comment:
Since we know interfaceList's length. You might want to define a bounded dest array.

Reply:
Fixed